File: huongB_devirt19/_slot16_dump.py
# SM3 raw-block DUMP (diagnose why slot16 reconstruction got 0). Hook 0xa0748, on each onEnter emit
#   {st_in(32B), inp(64B)}; force ONE device_register sign; reconstruct hashes in Python (split on IV).
import sys, os, time, frida
try: sys.stdout.reconfigure(encoding="utf-8", errors="replace")
except Exception: pass
HOST=os.environ.get("FRIDA_HOST","127.0.0.1:47119"); SIGN_OFF=os.environ.get("MS_SIGN_OFF","0x9ecc0")
PID=int(os.environ.get("MS_PID","15803"))
IV_LE='6f168073b9b21449d742241700068adabc306fa9aa3831164dee8de34e0efbb0'

JS=r"""
const LIB='libmetasec_ov.so'; const SIGN_OFF=%s; const SM3=0xa0748;
let mm=Process.findModuleByName(LIB); let base=mm?mm.base:null; let sign=null;
function initSign(){ if(!mm) return false; sign=new NativeFunction(base.add(SIGN_OFF),'pointer',['pointer','pointer']); return true; }
function hx(ab){ const u=new Uint8Array(ab); let s=''; for(let i=0;i<u.length;i++) s+=('0'+u[i].toString(16)).slice(-2); return s; }
send({t:'info',msg: mm?('base='+base):'no metasec'});
let CAP=false;
if(mm){ Interceptor.attach(base.add(SM3),{ onEnter(){
  if(!CAP) return; let st,inp;
  try{ st=hx(this.context.x0.add(8).readByteArray(32)); inp=hx(this.context.x1.readByteArray(64)); }catch(e){ return; }
  send({t:'blk', tid:this.threadId, st:st, inp:inp});
}});}
rpc.exports={ ready(){ return !!sign||initSign(); },
  cap(on){ CAP=!!on; },
  sign(url,hdr){ if(!sign&&!initSign()) throw new Error('no metasec');
    const u=Memory.allocUtf8String(url), h=Memory.allocUtf8String(hdr);
    const r=sign(u,h); return r.isNull()?null:r.readUtf8String(); } };
""" % SIGN_OFF

dev=frida.get_device_manager().add_remote_device(HOST); MSGS=[]
s=dev.attach(PID); sc=s.create_script(JS)
sc.on("message", lambda m,d: (MSGS.append(m["payload"]) if m.get("type")=="send" else print("[frida]",m)))
sc.load()
if not sc.exports_sync.ready(): print("[!] not ready"); sys.exit(1)
print("[*] attached pid=%d"%PID, flush=True)

TURL="https://api-boot.tiktokv.com/service/2/device_register/?device_platform=android&aid=1233&version_code=2024505040"
THDR=("x-ss-stub\r\n01205F31B47EC9C72AB1A5555960AA63\r\ncontent-type\r\napplication/json; charset=utf-8\r\n"
  "x-ss-req-ticket\r\n1756000000000\r\nsdk-version\r\n2\r\npassport-sdk-version\r\n1\r\n"
  "user-agent\r\ncom.zhiliaoapp.musically/2024505040")
sc.exports_sync.cap(True)
out=sc.exports_sync.sign(TURL, THDR)
time.sleep(0.4); sc.exports_sync.cap(False)
xa=""
for i,k in enumerate(out.replace("\r\n","\n").split("\n")[:-1]):
    if k.strip().lower()=="x-argus": xa=out.replace("\r\n","\n").split("\n")[i+1].strip()
blks=[m for m in MSGS if m.get("t")=="blk"]
print("[*] X-Argus len=%d  SM3 onEnter blocks=%d"%(len(xa), len(blks)), flush=True)

# group into hashes: a new hash begins where st==IV_LE
hashes=[]; cur=None
for b in blks:
    if b["st"]==IV_LE:
        if cur: hashes.append(cur)
        cur=[b["inp"]]
    elif cur is not None:
        cur.append(b["inp"])
if cur: hashes.append(cur)
print("[*] reconstructed %d hash(es):"%len(hashes))
for hi,h in enumerate(hashes):
    raw=bytes.fromhex("".join(h))            # padded message (blocks*64)
    L=len(raw)
    bitlen=int.from_bytes(raw[-8:],"big"); mlen=bitlen//8
    ok = 0<mlen<L and raw[mlen]==0x80
    msg=raw[:mlen] if ok else b""
    txt="".join(chr(x) if 32<=x<127 else "." for x in msg[:80])
    tail="".join(chr(x) if 32<=x<127 else "." for x in msg[-24:]) if ok else ""
    hasdp = b"device_platform=" in msg
    print("  #%d blocks=%d padL=%d mlen=%s lastbyte=%s dp=%s"%(hi,len(h),L, mlen if ok else "?", ("%02x"%msg[-1]) if ok else "?", hasdp))
    print("      head: %s"%txt)
    if ok: print("      tail: ...%s"%tail)
    if ok and hasdp and msg[-1]==0x30:
        slot=msg[mlen-17:mlen-1].hex()
        print("      >>> #19 CANDIDATE  slot16=%s"%slot)

File: huongB_devirt19/_slot16_dump.js
// _slot16_dump.js — detect nonzero slot16, scan for it, and dump a window around each hit
// in the persistent pool region to reveal the table entry structure (stride-32 observed).
// Goal: see whether a seed/metadata sits adjacent to each slot16 => direct (seed->slot16) pairs.
'use strict';
const SO='libmetasec_ov.so', SM3=0xa0748;
const IV='6f168073b9b21449d742241700068adabc306fa9aa3831164dee8de34e0efbb0';
function hx(ab){const u=new Uint8Array(ab);let s='';for(let i=0;i<u.length;i++)s+=('0'+u[i].toString(16)).slice(-2);return s;}
function rd(a,n){try{return hx(ptr(a).readByteArray(n));}catch(e){return null;}}
function region(a){
  try{ const m=Process.findModuleByAddress(a); if(m) return m.name+'+0x'+a.sub(m.base).toString(16); }catch(e){}
  try{ const r=Process.findRangeByAddress(a); if(r) return (r.file?r.file.path:'[anon]')+' '+r.protection; }catch(e){}
  return 'anon?';
}
function scanFor(hexpat){
  const patStr=hexpat.match(/../g).join(' ');
  const ranges=Process.enumerateRanges('rw-'); const hits=[];
  for(const r of ranges){ if(r.size>32*1024*1024) continue;
    try{ const found=Memory.scanSync(r.base,r.size,patStr);
      for(const f of found){ hits.push(f.address); if(hits.length>=12) break; } }catch(e){}
    if(hits.length>=12) break; }
  return hits;
}
function install(){
  const m=Process.findModuleByName(SO); if(!m) return false;
  const base=m.base; const chain={}; const seen={}; let done=0;
  Interceptor.attach(base.add(SM3),{onEnter(){
    if(done>=3) return;
    const tid=this.threadId; let st,inp;
    try{ st=hx(this.context.x0.add(8).readByteArray(32)); inp=new Uint8Array(this.context.x1.readByteArray(64)); }catch(e){ return; }
    if(st===IV) chain[tid]=Array.from(inp); else if(chain[tid]){ for(let i=0;i<64;i++) chain[tid].push(inp[i]); } else return;
    const a=chain[tid],L=a.length; if(L<9) return;
    let bl=0; for(let i=L-8;i<L;i++) bl=bl*256+a[i]; const mlen=bl/8;
    if(!(mlen>16&&mlen<L)||a[mlen]!==0x80) return;
    if(a[mlen-1]!==0x30||mlen<40){ delete chain[tid]; return; }
    let slot=''; let pr=0; for(let i=mlen-17;i<mlen-1;i++){slot+=('0'+a[i].toString(16)).slice(-2); if(a[i]>=0x20&&a[i]<=0x7e)pr++;}
    if(slot==='00'.repeat(16)||pr>=12||seen[slot]){ delete chain[tid]; return; }
    seen[slot]=1; done++;
    const hits=scanFor(slot);
    const dumps=[];
    for(const h of hits){
      // dump 32 bytes before + 48 after (aligned window around a 32-byte entry table)
      const w=ptr(h).sub(0x30);
      dumps.push({hit:h.toString(), region:region(h), win_at:w.toString(), bytes:rd(w,0x80)});
    }
    send({t:'dump', slot16:slot, nhits:hits.length, dumps:dumps});
    delete chain[tid];
  }});
  send({t:'info',msg:'slot16-dump installed'});
  return true;
}
if(Process.findModuleByName(SO)) install();
else Interceptor.attach(Module.findGlobalExportByName('android_dlopen_ext'),{onEnter(a){try{this.p=a[0].readCString();}catch(e){}},onLeave(){if(this.p&&this.p.indexOf(SO)>=0)install();}});
